### PR TITLE
Ts/38933 dump when changing settings via api

### PR DESCRIPTION
**Pull request by karottenreibe**
Addresses issue [TS-38933](https://cqse.atlassian.net/browse/TS-38933)

- [x] Changes are tested adequately
- [x] Agent's README.md updated in case of user-visible changes
- [x] CHANGELOG.md updated
- [x] Present new features in [N&N](https://cqse.atlassian.net/l/cp/KHSr81m1)
- [x] [TGA Tutorial](https://docs.teamscale.com/tutorial/setting-up-tga-java) updated
- [x] [TIA Tutorial](https://docs.teamscale.com/tutorial/tia-java/) updated

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

File: agent/src/main/java/com/teamscale/jacoco/agent/Agent.java [hunk 1]
```diff
@@ -147,23 +147,30 @@ protected void prepareShutdown() {
...
+			return;
 		}
+
+		try (Stream<Path> stream = Files.list(directory)) {
+			if (stream.count() == 0) {
+				return;
+			}
+		}
+
+		Files.delete(directory);
```
Comment (DreierF) [inline agent/src/main/java/com/teamscale/jacoco/agent/Agent.java:166]: This keeps the directory now if it is empty and attempt to delete it if it is not empty, but the code before did the opposite (delete iff empty).

File: system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+		TeamscaleMockServer teamscaleMockServer = new TeamscaleMockServer(
+				SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads();
+
+		new SystemUnderTest().foo();
+		SystemTestUtils.changePartition(SystemTestUtils.AGENT_PORT, "some_other_value");
+
+		assertThat(teamscaleMockServer.uploadedReports).hasSize(1);
+		String report = teamscaleMockServer.uploadedReports.get(0).getReportString();
+		// ensure that only the foo() method was covered, not the bar() method
```
Comment (DreierF) [inline system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java:29]: Why would `bar` appear here? AFAICS it is never executed.

File: system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		System.setProperty("org.eclipse.jetty.util.log.class", "org.eclipse.jetty.util.log.StdErrLog");
+		System.setProperty("org.eclipse.jetty.LEVEL", "OFF");
+
+		TeamscaleMockServer teamscaleMockServer = new TeamscaleMockServer(
+				SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads();
+
+		new SystemUnderTest().foo();
+		SystemTestUtils.changePartition(SystemTestUtils.AGENT_PORT, "some_other_value");
+
+		assertThat(teamscaleMockServer.uploadedReports).hasSize(1);
```
Comment (DreierF) [inline system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java:27]: Should we check that the report was uploaded under the old partition name as expected?

**Review by DreierF — CHANGES_REQUESTED** [review]

File: system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java (hunk 2 above)
Comment (karottenreibe) [reply system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java:29]: it appears with `mi="2"`, i.e. all probes are missed in that line. So this, as the comment states, checks that the bar method was *not* covered. I tried to refactor the code to make this clearer. Let me know if it worked

File: system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java (hunk 3 above)
Comment (karottenreibe) [reply system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java:27]: good idea

File: agent/src/main/java/com/teamscale/jacoco/agent/Agent.java (hunk 1 above)
Comment (karottenreibe) [reply agent/src/main/java/com/teamscale/jacoco/agent/Agent.java:166]: whoops, good catch! thanks for the thorough review!

File: system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java (hunk 2 above)
Comment (DreierF) [reply system-tests/api-changing-settings-should-dump/src/test/java/com/teamscale/tia/client/ApiChangingSettingsShouldDumpSystemTest.java:29]: That's clear to me. I was just confused as to why we even check for that condition i.e. no matter if the dumping would be broken or not we would never expect bar to be covered so why do we check for that.

**Review by DreierF — APPROVED** [review]

**Pull request merged**